### DISCUSSION TITLE
Which categories should we allow tracks to be filtered on?

### DISCUSSION THREAD

**Discussion by ErikSchierboom** [exercism/v3#2277] [Q&A] [answered]
In v3, the page that lists all the tracks, we want to implement some way of filtering of the tracks. I've tried to come up with some categories to filter on:

Compiles to:
  - Binary
  - Bytecode
  - JavaScript

Runtime:
  - Android
  - Browser
  - BEAM (Erlang)
  - Common Language Runtime (.NET)
  - iOS
  - JVM (Java)
  - V8 (NodeJS)

Paradigm:
  - Declarative
  - Functional
  - Imperative
  - Logic
  - Object-oriented
  - Procedural

Typing:
  - Static
  - Dynamic
  - Strong
  - Weak

Domain:
  - Cross-platform
  - Embedded systems
  - Financial
  - Games
  - GUIs
  - Logic
  - Maths
  - Mobile
  - Scientific
  - Scripting
  - Web development

Family:
  - C-like
  - Lisp
  - ML
  - Java
  - JavaScript
  - SASL
  - sh

Note: I've picked some of the more obvious families, but I'm sure I've missed some.

We'd be really interested in hearing what you would like to filter on and what you think about the above list.

**Comment by ynfle** [comment]
I'd say. The specific typing options are too specific, beginners won't know what they mean. If the option is used, it suggest filtering between weak and strong typing.

Other options could package manager support, native or there are bunch of options like cargo for rust, metaprogramming abilities (although I don't know how many tracks are going to teach that 

Regarding paradigms, will they be able to filtered for multiple paradigms for a language that is multi-paradigm

**Reply by ErikSchierboom** [reply]
The aim is to support selecting multiple options.

**Comment by halfdan** [comment]
This may be a meta-question. How do people pick languages in general? Personally I mostly just differentiate between scripted/compiled and "web"/"system" language (which supports the "Domain" selector). It seems strange for this page to become a way of language discovery - I would expect people who find exercism already know what language they are keen to learn.

**Reply by bobahop** [reply]
You might like to partake in this discussion...[What meta info should we show on a track page](https://github.com/exercism/v3/discussions/2276)

**Comment by bobahop** [comment]
Maybe robotics for a domain? There is some overlap with embedded, but there are also some distinctions.

**Reply by BethanyG** [reply]
Agree wit robotics as a specific one.  Python has a distinct distro for embedded and micro-controller stuff (MicroPython) -- but then has other libs and tools for robotics and things like LIDAR and "self driving cars", etc. etc.

**Comment by bobahop** [comment]
More domain suggestions...

financial (e.g. C++)
scientific (e.g. Python)

Then there are the categories of [AI, ML, DL, NLP, etc.](https://sonix.ai/articles/difference-between-artificial-intelligence-machine-learning-and-natural-language-processing)

**Reply by BethanyG** [reply]
"Scientific" even feels like too big of a category -- especially given all the specializations.  The libraries and languages for astro physics are quite different from the ones that are used for the modeling of wave stress or wind shear.

"Maths" feels too big as well.  There are things like automated proofs .. but then there are all of the statistics and probability specialties.  And then the particular financial applications.

**Reply by ghost** [reply]
As a student learning introductory sub-disciplines in mathematics, I've often searched for practice exercises by the sub-discipline. e.g. the monty hall problem for statistics and probability.

**Comment by iHiD** [comment]
Also see https://github.com/exercism/v3/issues/2763

**Comment by wneumann** [comment]
I'm not sure iOS or Android belong in _Runtime_ any more than Linux, Windows, or macOS would. Which in not to say that supported OSs is a bad category to filter on, but it is different from runtimes supported.

You may want to consider adding server-side to your domains. Another filtering point might be whether or not it offers an interactive programming environment like a REPL or like Swift Playgrounds.

**Comment by ghost** [comment, marked as answer]
That list is a great start!

Who should the tags primarily serve? Absolute beginners? Programmers well versed in one language seeking to learn another using exercism.io?

This seems like a taxonomy problem to me. By identifying the key audience it would help constrain the relevant terms in the taxonomy.

Do track maintainers need to generate relevant terms?

As an aside: It would be interesting to attempt deriving the categories from all track's concept and story lists. Then we could take an intersection to find a broadly relevant set with minimal manual categorization.

**Reply by ErikSchierboom** [reply]
Well, no one should be an absolute beginner on exercism, as our target audience is people that are familiar with at least one programming language.

Track maintainers can't generate terms themselves, but instead choose from a predefined list of tags.

**Reply by iHiD** [reply]
We've now settled on an initial set of tags, which Erik has added to the docs.

PR here: https://github.com/exercism/v3-docs/pull/17#issuecomment-763165816
Docs here: https://github.com/exercism/v3-docs/blob/master/anatomy/tracks/config-json.md#tags

We'll "formally" ask everyone to assign these as part of our v3-launch work coming in the next few days.

Thanks everyone :)
